src/Raven.Server/Documents/ShardedTcpHandlers/ShardedSubscriptionConnection.cs

```diff
@@ -0,0 +1,328 @@
+                shardsToReconnect.Clear();
+                if (didBatch == false)
+                    await Task.Delay(500);
+
+                if (heartbeatSp.ElapsedMilliseconds >= 3000)
+                {
+                    await SendHeartBeat("Waited for 3000ms for batch from shard workers");
+                    heartbeatSp.Restart();
+                }
+                didBatch = false;
+                var stopping = false;
+                foreach ((string shard, SubscriptionShardHolder shardHolder) in _shardWorkers)
+                {
+                    CancellationTokenSource.Token.ThrowIfCancellationRequested();
+
+                    if (shardHolder.PullingTask.IsFaulted == false)
+                    {
+                        ShardedSubscriptionWorker.PublishedBatch batch = shardHolder.Worker.PublishedBatchItem;
+                        if (batch == null)
+                            continue;
+
+                        didBatch = true;
+                        // Send to client
+                        try
+                        {
+                            await WriteBatchToClientAndAck(batch, shard);
+                            batch.SendBatchToClientTcs.SetResult();
+                        }
+                        catch (Exception e)
+                        {
+                            // need to fail the shard subscription worker
+                            batch.SendBatchToClientTcs.TrySetException(e);
+                            throw;
+                        }
+                        await batch.ConfirmFromShardSubscriptionConnectionTcs.Task;
+
+                        // send confirm to client and continue processing
+                        await SendConfirmToClient();
+                        continue;
+                    }
+
+                    exceptions.Add(shard, shardHolder.PullingTask.Exception);
+
+                    if (stopping)
+                        continue;
+
+                    shardsToReconnect.Add(shard);
+
+                    if (shardHolder.LastErrorDateTime.HasValue == false)
+                    {
+                        shardHolder.LastErrorDateTime = DateTime.UtcNow;
+                        continue;
+                    }
+
+                    if (DateTime.UtcNow - shardHolder.LastErrorDateTime.Value <= _options.ShardedConnectionMaxErroneousPeriod)
+                        continue;
+
+                    // we are stopping this subscription
+                    stopping = true;
+                }
+
+                if (exceptions.Count == _shardWorkers.Count && stopping == false)
+                {
+                    stopping = true;
+                    foreach (var worker in _shardWorkers)
+                    {
+                        if (exceptions.TryGetValue(worker.Key, out var ex) == false)
+                            throw new InvalidOperationException($"Expected to get exception for shard worker {worker.Key} but could not find.");
```

Comment:
Why do you care about this here?

Reply:
it should not happen, I changed to `Debug.Assert`